File: lib/api/routes/otherBilan.js
//Presente les autres methodes que je pouvait utiliser

//un exemple (on ne l'a pas utiliser c'est pour juste voir la maniere de proceder elle ne donne pas le resultat exact)
router.post('/agentTopBroyage', async (req, res) => {
    try {
        let { heureDebut, heureFin } = req.body;

        // Conversion des heures de chaînes en nombres entiers
        heureDebut = parseInt(heureDebut, 10);
        heureFin = parseInt(heureFin, 10);

        // Validation des paramètres après conversion
        if (isNaN(heureDebut) || isNaN(heureFin) || heureDebut < 0 || heureDebut > 23 || heureFin < 0 || heureFin > 23) {
            return res.status(400).json({ error: 'heureDebut et heureFin doivent être des nombres entre 0 et 23' });
        }

        const pool = await poolPromise;
        const dateNow = new Date();  // Date actuelle

        // Construction de la requête SQL
        let query = `
            SELECT TOP 1
                agentBroyage AS matricule,
                SUM(poids) AS totalPoids
            FROM TAS
            WHERE 
                etat = 1
                AND heureBroyage BETWEEN @heureDebut AND @heureFin
                AND dateBroyage = CAST(@dateNow AS DATE)
            GROUP BY agentBroyage
            ORDER BY totalPoids DESC;
        `;

        const result = await pool.request()
            .input('dateNow', sql.DateTime, dateNow)
            .input('heureDebut', sql.Int, heureDebut)
            .input('heureFin', sql.Int, heureFin)
            .query(query);

        if (result.recordset.length === 0) {
            return res.status(404).json({ error: 'Aucun tas broyé dans la plage horaire spécifiée' });
        }

        // Récupération de l'agent ayant broyé le plus
        const topAgent = result.recordset[0];

        res.json({
            matricule: topAgent.matricule,
            totalPoids: topAgent.totalPoids
        });
    } catch (err) {
        console.error('SQL error:', err.message);
        res.status(500).json({ error: 'Internal Server Error' });
    }
});

//un exemple (on ne l'a pas utiliser c'est pour juste voir la maniere de proceder elle ne donne pas le resultat exact)
router.post('/agentTopBroyage', async (req, res) => {
    try {
        let { heureDebut, heureFin } = req.body;

        // Conversion des heures de chaînes en nombres entiers
        heureDebut = parseInt(heureDebut, 10);
        heureFin = parseInt(heureFin, 10);

        // Validation des paramètres après conversion
        if (isNaN(heureDebut) || isNaN(heureFin) || heureDebut < 0 || heureDebut > 23 || heureFin < 0 || heureFin > 23) {
            return res.status(400).json({ error: 'heureDebut et heureFin doivent être des nombres entre 0 et 23' });
        }

        const pool = await poolPromise;
        const dateNow = new Date(); // Date actuelle

        // Requête SQL pour obtenir la somme des poids broyés dans TAS et DECHARGERTABLE
        const query = `
            SELECT TOP 1
                agentBroyage AS matricule,
                (SUM(T.poids) + ISNULL(SUM(D.poidsNet), 0)) AS totalPoids
            FROM TAS T
            LEFT JOIN DECHARGERTABLE D ON T.agentBroyage = D.agentMatricule 
            WHERE 
                T.etat = 1
                AND T.heureBroyage BETWEEN @heureDebut AND @heureFin
                AND T.dateBroyage = CAST(@dateNow AS DATE)
                AND (D.dateHeureDecharg IS NULL OR (D.dateHeureDecharg BETWEEN @heureDebut AND @heureFin))
            GROUP BY T.agentBroyage
            ORDER BY totalPoids DESC;
        `;

        const result = await pool.request()
            .input('dateNow', sql.DateTime, dateNow)
            .input('heureDebut', sql.Int, heureDebut)
            .input('heureFin', sql.Int, heureFin)
            .query(query);

        if (result.recordset.length === 0) {
            return res.status(404).json({ error: 'Aucun agent n\'a broyé de tas ou déchargé de camions dans la plage horaire spécifiée' });
        }

        // Récupération de l'agent ayant broyé le plus
        const topAgent = result.recordset[0];

        res.json({
            matricule: topAgent.matricule,
            totalPoids: topAgent.totalPoids
        });
    } catch (err) {
        console.error('SQL error:', err.message);
        res.status(500).json({ error: 'Internal Server Error' });
    }
});

//un exemple (on ne l'a pas utiliser c'est pour juste voir la maniere de proceder elle ne donne pas le resultat exact)
router.post('/agentTopBroyage', async (req, res) => {
    try {
        let { heureDebut, heureFin } = req.body;

        // Conversion des heures de chaînes en nombres entiers
        heureDebut = parseInt(heureDebut, 10);
        heureFin = parseInt(heureFin, 10);

        // Validation des paramètres après conversion
        if (isNaN(heureDebut) || isNaN(heureFin) || heureDebut < 0 || heureDebut > 23 || heureFin < 0 || heureFin > 23) {
            return res.status(400).json({ error: 'heureDebut et heureFin doivent être des nombres entre 0 et 23' });
        }

        const pool = await poolPromise;
        const dateNow = new Date(); // Date actuelle

        // Requête SQL pour obtenir la somme des poids broyés dans TAS et DECHARGERTABLE
        const query = `
            SELECT TOP 1
                agentBroyage AS matricule,
                (SUM(T.poids) + ISNULL(SUM(D.poidsNet), 0)) AS totalPoids
            FROM TAS T
            LEFT JOIN DECHARGERTABLE D ON T.agentBroyage = D.agentMatricule 
            WHERE 
                T.etat = 1
                AND T.heureBroyage BETWEEN @heureDebut AND @heureFin
                AND T.dateBroyage = CAST(@dateNow AS DATE)
                AND (D.dateHeureDecharg IS NULL OR (D.dateHeureDecharg BETWEEN @heureDebut AND @heureFin))
            GROUP BY T.agentBroyage
            ORDER BY totalPoids DESC;
        `;

        const result = await pool.request()
            .input('dateNow', sql.DateTime, dateNow)
            .input('heureDebut', sql.Int, heureDebut)
            .input('heureFin', sql.Int, heureFin)
            .query(query);

        if (result.recordset.length === 0) {
            return res.status(404).json({ error: 'Aucun agent n\'a broyé de tas ou déchargé de camions dans la plage horaire spécifiée' });
        }

        // Récupération de l'agent ayant broyé le plus
        const topAgent = result.recordset[0];

        res.json({
            matricule: topAgent.matricule,
            totalPoids: topAgent.totalPoids
        });
    } catch (err) {
        console.error('SQL error:', err.message);
        res.status(500).json({ error: 'Internal Server Error' });
    }
});

//un exemple (on ne l'a pas utiliser c'est pour juste voir la maniere de proceder elle ne donne pas le resultat exact)
router.post('/agentMinBroyage', async (req, res) => {
    try {
        const { heureDebut, heureFin } = req.body;

        // Validation des paramètres
        if (typeof heureDebut !== 'number' || typeof heureFin !== 'number' || heureDebut < 0 || heureDebut > 23 || heureFin < 0 || heureFin > 23) {
            return res.status(400).json({ error: 'heureDebut et heureFin doivent être des nombres entre 0 et 23' });
        }

        const pool = await poolPromise;
        const dateNow = new Date();

        let query = '';

        if (heureDebut <= heureFin) {
            // Même journée
            query = `
                SELECT agentBroyage, SUM(poids) AS totalPoids
                FROM TAS
                WHERE etat = 0
                AND CAST(dateBroyage AS DATE) = CAST(@dateNow AS DATE)
                AND heureBroyage BETWEEN @heureDebut AND @heureFin
                GROUP BY agentBroyage
                ORDER BY totalPoids ASC
                OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY;
            `;
        } else {
            // Débordement sur le jour suivant
            query = `
                SELECT agentBroyage, SUM(poids) AS totalPoids
                FROM TAS
                WHERE etat = 0
                AND (
                    (CAST(dateBroyage AS DATE) = CAST(@dateNow AS DATE) AND heureBroyage >= @heureDebut)
                    OR
                    (CAST(dateBroyage AS DATE) = DATEADD(DAY, 1, @dateNow) AND heureBroyage <= @heureFin)
                )
                GROUP BY agentBroyage
                ORDER BY totalPoids ASC
                OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY;
            `;
        }

        const result = await pool.request()
            .input('dateNow', sql.DateTime, dateNow)
            .input('heureDebut', sql.Int, heureDebut)
            .input('heureFin', sql.Int, heureFin)
            .query(query);

        if (result.recordset.length > 0) {
            const agentMinBroyage = result.recordset[0];
            res.json({
                matricule: agentMinBroyage.agentBroyage,
                totalPoids: agentMinBroyage.totalPoids
            });
        } else {
            res.status(404).json({ error: 'Aucun agent n\'a broyé des tas dans cette plage horaire' });
        }
    } catch (err) {
        console.error('SQL error:', err.message);
        res.status(500).json({ error: 'Internal Server Error' });
    }
});


//(peut etre utiliser)
// Route pour récupérer le tonnage total de canne entrée entre deux dates
router.get('/getTonnageCanneEntreeEntreDates', async (req, res) => {
    try {
        const { dateDebut, dateFin } = req.query;

        if (!dateDebut || !dateFin) {
            return res.status(400).json({ error: 'Les paramètres dateDebut et dateFin sont requis' });
        }

        const pool = await poolPromise;

        // Requête pour récupérer la somme des poids nets dans DECHARGERCOURS entre les deux dates
        const queryCours = `
            SELECT SUM(poidsNet) AS totalPoidsNetCours
            FROM DECHARGERCOURS
            WHERE dateHeureDecharg BETWEEN @dateDebut AND @dateFin;
        `;

        // Requête pour récupérer la somme des poids nets dans DECHARGERTABLE entre les deux dates
        const queryTable = `
            SELECT SUM(poidsNet) AS totalPoidsNetTable
            FROM DECHARGERTABLE
            WHERE dateHeureDecharg BETWEEN @dateDebut AND @dateFin;
        `;

        // Exécuter les deux requêtes en parallèle
        const [resultCours, resultTable] = await Promise.all([
            pool.request()
                .input('dateDebut', sql.DateTime, new Date(dateDebut))
                .input('dateFin', sql.DateTime, new Date(dateFin))
                .query(queryCours),
            pool.request()
                .input('dateDebut', sql.DateTime, new Date(dateDebut))
                .input('dateFin', sql.DateTime, new Date(dateFin))
                .query(queryTable),
        ]);

        const totalPoidsNetCours = resultCours.recordset[0].totalPoidsNetCours || 0;
        const totalPoidsNetTable = resultTable.recordset[0].totalPoidsNetTable || 0;

        // Calculer le tonnage total
        const tonnageTotal = totalPoidsNetCours + totalPoidsNetTable;

        // Retourner le résultat en JSON
        res.json({ tonnageTotal });
    } catch (error) {
        console.error('Erreur lors de la récupération du tonnage de canne entrée entre deux dates:', error);
        res.status(500).json({ error: 'Erreur interne du serveur' });
    }
});

//(peut etre utiliser)
// Route pour obtenir l'agent qui a broyé le plus entre deux dates
router.post('/agentTopBroyageEntreDates', async (req, res) => {
    try {
        let { dateDebut, dateFin } = req.body;

        // Validation des dates
        if (!dateDebut || !dateFin) {
            return res.status(400).json({ error: 'Les paramètres dateDebut et dateFin sont requis' });
        }

        const pool = await poolPromise;

        // Requête SQL pour obtenir la somme des poids broyés dans TAS et DECHARGERTABLE
        const query = `
            SELECT TOP 1
                agentBroyage AS matricule,
                (SUM(T.poids) + ISNULL(SUM(D.poidsNet), 0)) AS totalPoids
            FROM TAS T
            LEFT JOIN DECHARGERTABLE D ON T.agentBroyage = D.agentMatricule 
            WHERE 
                T.etat = 1
                AND T.dateBroyage BETWEEN @dateDebut AND @dateFin
                AND (D.dateHeureDecharg IS NULL OR D.dateHeureDecharg BETWEEN @dateDebut AND @dateFin)
            GROUP BY T.agentBroyage
            ORDER BY totalPoids DESC;
        `;

        const result = await pool.request()
            .input('dateDebut', sql.DateTime, new Date(dateDebut))
            .input('dateFin', sql.DateTime, new Date(dateFin))
            .query(query);

        if (result.recordset.length === 0) {
            return res.status(404).json({ error: 'Aucun agent n\'a broyé de tas ou déchargé de camions dans la plage de dates spécifiée' });
        }

        // Récupération de l'agent ayant broyé le plus
        const topAgent = result.recordset[0];

        res.json({
            matricule: topAgent.matricule,
            totalPoids: topAgent.totalPoids
        });
    } catch (err) {
        console.error('SQL error:', err.message);
        res.status(500).json({ error: 'Erreur interne du serveur' });
    }
});

//(peut etre utiliser)
// Endpoint pour récupérer le tonnage broyé directement à partir de la table DECHARGERTABLE Entre deux dates
router.post('/tonnageBroyerDirectDates', async (req, res) => {
    try {
        let { dateDebut, dateFin } = req.body;

        // Validation des dates
        if (!dateDebut || !dateFin) {
            return res.status(400).json({ error: 'dateDebut et dateFin sont requis' });
        }

        const pool = await poolPromise;

        const query = `
            SELECT SUM(poidsNet) AS totalPoidsNet
            FROM DECHARGERTABLE
            WHERE 
            CAST(dateHeureDecharg AS DATE) BETWEEN @dateDebut AND @dateFin;
        `;

        const result = await pool.request()
            .input('dateDebut', sql.Date, dateDebut)
            .input('dateFin', sql.Date, dateFin)
            .query(query);

        const stockBroyerDirect = result.recordset[0].totalPoidsNet || 0;

        res.json({ stockBroyerDirect });
    } catch (err) {
        console.error('SQL error:', err.message);
        res.status(500).json({ error: 'Internal Server Error' });
    }
});

//(peut etre utiliser)
// Endpoint pour récupérer le tonnage broyé par tas entre deux dates
router.post('/tonnageBroyerParTasDates', async (req, res) => {
    try {
        let { dateDebut, dateFin } = req.body;

        // Validation des dates
        if (!dateDebut || !dateFin) {
            return res.status(400).json({ error: 'dateDebut et dateFin sont requis' });
        }

        const pool = await poolPromise;

        const query = `
            SELECT SUM(poids) AS totalPoidsTas
            FROM TAS
            WHERE 
            etat = 1  -- Le tas est broyé (etat = 1)
            AND CAST(dateBroyage AS DATE) BETWEEN @dateDebut AND @dateFin;
        `;

        const result = await pool.request()
            .input('dateDebut', sql.Date, dateDebut)
            .input('dateFin', sql.Date, dateFin)
            .query(query);

        const tonnageBroyerParTas = result.recordset[0].totalPoidsTas || 0;

        res.json({ tonnageBroyerParTas });
    } catch (err) {
        console.error('SQL error:', err.message);
        res.status(500).json({ error: 'Internal Server Error' });
    }
});


// Route pour récupérer le tonnage total de canne entrée pour l'année en cours 
router.get('/getTonnageCanneEntree', async (req, res) => {
    try {
        const pool = await poolPromise;
        const currentYear = new Date().getFullYear();

        // Requête pour récupérer la somme des poids nets dans DECHARGERCOURS pour l'année en cours
        const queryCours = `
            SELECT SUM(poidsNet) AS totalPoidsNetCours
            FROM DECHARGERCOURS
            WHERE YEAR(dateHeureDecharg) = @currentYear;
        `;

        // Requête pour récupérer la somme des poids nets dans DECHARGERTABLE pour l'année en cours
        const queryTable = `
            SELECT SUM(poidsNet) AS totalPoidsNetTable
            FROM DECHARGERTABLE
            WHERE YEAR(dateHeureDecharg) = @currentYear;
        `;

        // Exécuter les deux requêtes en parallèle
        const [resultCours, resultTable] = await Promise.all([
            pool.request().input('currentYear', sql.Int, currentYear).query(queryCours),
            pool.request().input('currentYear', sql.Int, currentYear).query(queryTable),
        ]);

        const totalPoidsNetCours = resultCours.recordset[0].totalPoidsNetCours || 0;
        const totalPoidsNetTable = resultTable.recordset[0].totalPoidsNetTable || 0;

        // Calculer le tonnage total
        const tonnageTotal = totalPoidsNetCours + totalPoidsNetTable;

        // Retourner le résultat en JSON
        res.json({ tonnageTotal });
    } catch (error) {
        console.error('Erreur lors de la récupération du tonnage de canne entrée:', error);
        res.status(500).json({ error: 'Erreur interne du serveur' });
    }
});


// Route pour récupérer le tonnage total de canne entrée pour le jour (on respecte les contraintes de quart)
router.get('/getTonnageCanneEntree', async (req, res) => {
    try {
        const pool = await poolPromise;

        // Calculer la date et l'heure actuelles
        const maintenant = new Date();

        // Déterminer le début de la période à 06:00 aujourd'hui
        let debutPeriode = new Date(maintenant);
        debutPeriode.setHours(6, 0, 0, 0); // Début de la période à 06:00

        // Si l'heure actuelle est avant 06:00, ajuster le début de la période à 06:00 hier
        if (maintenant < debutPeriode) {
            debutPeriode.setDate(debutPeriode.getDate() - 1); // Commencer à 06:00 le jour précédent
        }

        // Déterminer la fin de la période à 06:00 le lendemain
        const finPeriode = new Date(debutPeriode);
        finPeriode.setDate(finPeriode.getDate() + 1); // Fin de la période à 06:00 le jour suivant

        // Requête pour récupérer la somme des poids nets dans DECHARGERCOURS
        const requeteCours = `
            SELECT SUM(poidsNet) AS totalPoidsNetCours
            FROM DECHARGERCOURS
            WHERE dateHeureDecharg >= @debutPeriode AND dateHeureDecharg < @finPeriode;
        `;

        // Requête pour récupérer la somme des poids nets dans DECHARGERTABLE
        const requeteTable = `
            SELECT SUM(poidsNet) AS totalPoidsNetTable
            FROM DECHARGERTABLE
            WHERE dateHeureDecharg >= @debutPeriode AND dateHeureDecharg < @finPeriode;
        `;

        // Exécuter les deux requêtes en parallèle
        const [resultatCours, resultatTable] = await Promise.all([
            pool.request()
                .input('debutPeriode', sql.DateTime, debutPeriode)
                .input('finPeriode', sql.DateTime, finPeriode)
                .query(requeteCours),
            pool.request()
                .input('debutPeriode', sql.DateTime, debutPeriode)
                .input('finPeriode', sql.DateTime, finPeriode)
                .query(requeteTable),
        ]);

        // Extraire les résultats des requêtes
        const totalPoidsNetCours = resultatCours.recordset[0].totalPoidsNetCours || 0;
        const totalPoidsNetTable = resultatTable.recordset[0].totalPoidsNetTable || 0;

        // Calculer le tonnage total
        const tonnageTotal = totalPoidsNetCours + totalPoidsNetTable;

        // Retourner le résultat en JSON
        res.json({ tonnageTotal });
    } catch (error) {
        console.error('Erreur lors de la récupération du tonnage de canne entrée:', error);
        res.status(500).json({ error: 'Erreur interne du serveur' });
    }
});

File: lib/services/api/other_bilan.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:logger/logger.dart';

final logger = Logger();

//autres bilan qu'on pourait avoir

Future<void> main() async {

final maxBroyageHoraire = await getTopAgentFromAPI(heureDebut: 19, heureFin: 20);
final minBroyageHoraire = await getMinAgentFromAPI(heureDebut: 19, heureFin: 20);
final maxBroyageAnnee = await getTopAgentAnneeFromAPI();
final minBroyageAnnee = await getMinAgentAnneeFromAPI();


if (maxBroyageHoraire == null) {
  logger.e('Une erreur est survenue lors de la requête.');
} else if (maxBroyageHoraire.isEmpty) {
  logger.i('Aucun agent n\'a été trouvé pour la plage horaire spécifiée.');
} else {
  logger.d('Matricule: ${maxBroyageHoraire['matricule']}');
  logger.d('max Broyage Horaire: ${maxBroyageHoraire['totalPoids']}');
}


if (minBroyageHoraire == null) {
  logger.e('Une erreur est survenue lors de la requête.');
} else if (minBroyageHoraire.isEmpty) {
  logger.i('Aucun agent n\'a été trouvé pour la plage horaire spécifiée.');
} else {
  logger.d('Matricule: ${minBroyageHoraire['matricule']}');
  logger.d('min Broyage Horaire: ${minBroyageHoraire['totalPoids']}');
}


if (maxBroyageAnnee == null) {
  logger.e('Une erreur est survenue lors de la requête.');
} else if (maxBroyageAnnee.isEmpty) {
  logger.i('Aucun agent n\'a été trouvé pour cette annee.');
} else {
  logger.d('Matricule: ${maxBroyageAnnee['matricule']}');
  logger.d('max Broyage Annee: ${maxBroyageAnnee['totalPoids']}');
}


if (minBroyageAnnee == null) {
  logger.e('Une erreur est survenue lors de la requête.');
} else if (minBroyageAnnee.isEmpty) {
  logger.i('Aucun agent n\'a été trouvé pour cette annee.');
} else {
  logger.d('Matricule: ${minBroyageAnnee['matricule']}');
  logger.d('min Broyage Annee: ${minBroyageAnnee['totalPoids']}');
}

// Exemple d'appel avec deux dates
final dateDebut = DateTime(2024, 1, 1); // Exemple: 1er Janvier 2024
final dateFin = DateTime(2024, 12, 31); // Exemple: 31 Décembre 2024

afficherTonnageCanneEntreeEntreDates(dateDebut, dateFin);

afficherAgentTopBroyage(dateDebut, dateFin);

}


void afficherTonnageCanneEntreeEntreDates(DateTime dateDebut, DateTime dateFin) async {
  try {
    final tonnageTotal = await getTonnageCanneEntreeEntreDates(
      dateDebut: dateDebut,
      dateFin: dateFin,
    );
    logger.d('Le tonnage total de canne entrée entre ${dateDebut.toIso8601String()} et ${dateFin.toIso8601String()} est de : $tonnageTotal tonnes.');
  } catch (e) {
    logger.e('Erreur lors de la récupération du tonnage : $e');
  }
}


void afficherAgentTopBroyage(DateTime dateDebut, DateTime dateFin) async {
  try {
    final agentTopBroyage = await getAgentTopBroyageEntreDates(
      dateDebut: dateDebut,
      dateFin: dateFin,
    );
    logger.d('L\'agent avec le plus de broyage entre ${dateDebut.toIso8601String()} et ${dateFin.toIso8601String()} est : ${agentTopBroyage['matricule']} avec ${agentTopBroyage['totalPoids']} tonnes.');
  } catch (e) {
    logger.e('Erreur lors de la récupération de l\'agent top broyage : $e');
  }
}


//Fonction pour obtenir les tonnages broyés directement et par tas entre les dates fournies.
void afficherTonnageBroyerEntreDates(DateTime dateDebut, DateTime dateFin) async {
  try {
    final tonnageBroyerDirect = await getTonnageBroyerDirectEntreDates(
      dateDebut: dateDebut,
      dateFin: dateFin,
    );

    final tonnageBroyerParTas = await getTonnageBroyerParTasEntreDates(
      dateDebut: dateDebut,
      dateFin: dateFin,
    );

    logger.d('Tonnage broyé entre les dates : Direct = $tonnageBroyerDirect tonnes, Par Tas = $tonnageBroyerParTas tonnes');
  } catch (e) {
    logger.e('Erreur lors de la récupération du tonnage broyé : $e');
  }
}


// Fonction pour récupérer l'agent qui a broyé le plus entre deux dates
Future<Map<String, dynamic>> getAgentTopBroyageEntreDates({
  required DateTime dateDebut,
  required DateTime dateFin,
}) async {
  final uri = Uri.parse('http://192.168.1.190:80/api/agentTopBroyageEntreDates');
  
  final response = await http.post(
    uri,
    headers: {"Content-Type": "application/json"},
    body: json.encode({
      'dateDebut': dateDebut.toIso8601String(),
      'dateFin': dateFin.toIso8601String(),
    }),
  );

  if (response.statusCode == 200) {
    return json.decode(response.body);
  } else {
    throw Exception('Erreur lors de la récupération de l\'agent top broyage');
  }
}


// Fonction pour récupérer le tonnage total de canne entrée entre deux dates
Future<double> getTonnageCanneEntreeEntreDates({
  required DateTime dateDebut,
  required DateTime dateFin,
}) async {
  final uri = Uri.parse(
      'http://192.168.1.190:80/api/getTonnageCanneEntreeEntreDates?dateDebut=${dateDebut.toIso8601String()}&dateFin=${dateFin.toIso8601String()}');

  try {
    final response = await http.get(uri);

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['tonnageTotal'].toDouble();
    } else {
      throw Exception('Erreur lors de la récupération du tonnage de canne');
    }
  } catch (e) {
    logger.e('Erreur lors de la récupération du tonnage de canne: $e');
    throw Exception('Erreur lors de la récupération du tonnage de canne $e');
  }
}

//Fonction pour récupérer le tonnage broyé directement entre deux dates
Future<double> getTonnageBroyerDirectEntreDates({
  required DateTime dateDebut,
  required DateTime dateFin,
}) async {
  final uri = Uri.parse(
    'http://10.0.2.2:3000/tonnageBroyerDirectDates',
  );

  try {
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'dateDebut': dateDebut.toIso8601String(),
        'dateFin': dateFin.toIso8601String(),
      }),
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['stockBroyerDirect'].toDouble();
    } else {
      throw Exception('Erreur lors de la récupération du tonnage broyé directement');
    }
  } catch (e) {
    logger.e('Erreur lors de la récupération du tonnage broyé directement: $e');

    throw Exception('Erreur lors de la récupération du tonnage broyé directement $e');
  }
}


//Fonction pour récupérer le tonnage broyé par tas entre deux dates
Future<double> getTonnageBroyerParTasEntreDates({
  required DateTime dateDebut,
  required DateTime dateFin,
}) async {
  final uri = Uri.parse(
    'http://10.0.2.2:3000/tonnageBroyerParTasDates',
  );

  try {
    final response = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'dateDebut': dateDebut.toIso8601String(),
        'dateFin': dateFin.toIso8601String(),
      }),
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['tonnageBroyerParTas'].toDouble();
    } else {
      throw Exception('Erreur lors de la récupération du tonnage broyé par tas');
    }
  } catch (e) {
    logger.e('Erreur lors de la récupération du tonnage broyé par tas: $e');
    throw Exception('Erreur lors de la récupération du tonnage broyé par tas $e');

  }
}


// Fonction pour récupérer le tonnage total de canne entrée pour l'année en cours
Future<double?> getTonnageCanneEntree() async {
  final uri = Uri.parse('http://192.168.1.190:80/api/getTonnageCanneEntree');

  try {
    final response = await http.get(uri);

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return data['tonnageTotal'].toDouble();
    } else {
      //throw Exception('Erreur lors de la récupération du tonnage de canne');
      return null;
    }
  } catch (e) {
    logger.e('Erreur lors de la récupération du tonnage de canne: $e');
    return null;
  }
}


//Fonction d'appel pour l'agent avec le plus de tonnage broyé dans une plage horaire
Future<Map<String, dynamic>?> getTopAgentFromAPI({
  required int heureDebut, // Heure de début à envoyer
  required int heureFin,   // Heure de fin à envoyer
}) async {
  final uri = Uri.parse('http://192.168.1.190:80/api/agentTopBroyage'); // URL de l'API

  try {
    final response = await http.post(
      uri,
      headers: {
        'Content-Type': 'application/json',
      },
      body: json.encode({
        'heureDebut': heureDebut, // Envoi des heures de début
        'heureFin': heureFin,     // Envoi des heures de fin
      }),
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return {
        'matricule': data['matricule'],
        'totalPoids': data['totalPoids'].toDouble(),
      };
    } else if (response.statusCode == 404) {
      // Gérer le cas où aucun agent n'a été trouvé
      //logger.d('Aucun agent n\'a broyé de tas ou déchargé de camions dans la plage horaire spécifiée');
      return {};
    } else {
      // Gérer d'autres erreurs HTTP
      //logger.e('Erreur HTTP: ${response.statusCode}');
      return null;
    }
  } catch (e) {
    // Gérer les erreurs de requête
    //logger.e('Erreur lors de la requête: $e');
    return null;
  }
}


// recuperer l'agent qui broyer le moins de canne dans une plage horaire donnee 
Future<Map<String, dynamic>?> getMinAgentFromAPI({
  required int heureDebut, // Heure de début à envoyer
  required int heureFin,   // Heure de fin à envoyer
}) async {
  final uri = Uri.parse('http://192.168.1.190:80/api/agentMinBroyage'); // URL de l'API

  try {
    final response = await http.post(
      uri,
      headers: {
        'Content-Type': 'application/json',
      },
      body: json.encode({
        'heureDebut': heureDebut, // Envoi des heures de début
        'heureFin': heureFin,     // Envoi des heures de fin
      }),
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return {
        'matricule': data['matricule'],
        'totalPoids': data['totalPoids'].toDouble(),
      };
    } else if (response.statusCode == 404) {
      // Gérer le cas où aucun agent n'a broyé de tas
      //logger.d('Aucun agent n\'a broyé de tas ou déchargé de camions dans la plage horaire spécifiée');
      return {};
    } else {
      // Gérer d'autres erreurs HTTP
      //logger.e('Erreur HTTP: ${response.statusCode}');
      return null;
    }
  } catch (e) {
    // Gérer les erreurs de requête
    //logger.e('Erreur lors de la requête: $e');
    return null;
  }
}


//Fonction d'appel pour l'agent avec le plus de tonnage broyé sur l'année  (une autre façon de faire ne donne pas le resultat exact c'est pour juste voir la maniere de faire)
Future<Map<String, dynamic>?> getTopAgentAnneeFromAPI() async {
  final uri = Uri.parse('http://192.168.1.190:80/api/agentTopBroyageAnnee'); // URL de l'API

  try {
    final response = await http.get(
      uri,
      headers: {
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return {
        'matricule': data['matricule'],
        'totalPoids': data['totalPoids'].toDouble(),
      };
    } else if (response.statusCode == 404) {
      // Gérer le cas où aucun agent n'a été trouvé
      //logger.d('Aucun agent n\'a broyé de tas ou déchargé de camions cette année.');
      return {};
    } else {
      // Gérer d'autres erreurs HTTP
      //logger.e('Erreur HTTP: ${response.statusCode}');
      return null;
    }
  } catch (e) {
    // Gérer les erreurs de requête
    //logger.e('Erreur lors de la requête: $e');
    return null;
  }
}


//Fonction d'appel pour l'agent avec le moins de tonnage broyé sur l'année   (une autre façon de faire ne donne pas le resultat exact c'est pour juste voir la maniere de faire)
Future<Map<String, dynamic>?> getMinAgentAnneeFromAPI() async {
  final uri = Uri.parse('http://192.168.1.190:80/api/agentMinBroyageAnnee'); // URL de l'API

  try {
    final response = await http.get(
      uri,
      headers: {
        'Content-Type': 'application/json',
      },
    );

    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return {
        'matricule': data['matricule'],
        'totalPoids': data['totalPoids'].toDouble(),
      };
    } else if (response.statusCode == 404) {
      // Gérer le cas où aucun agent n'a été trouvé
      //logger.d('Aucun agent n\'a broyé de tas ou déchargé de camions cette année.');
      return {};
    } else {
      // Gérer d'autres erreurs HTTP
      //logger.e('Erreur HTTP: ${response.statusCode}');
      return null;
    }
  } catch (e) {
    // Gérer les erreurs de requête
    //logger.e('Erreur lors de la requête: $e');
    return null;
  }
}

